kad-dht/README.md

```diff
@@ -0,0 +1,427 @@
+|-----------------|----------------|--------|-----------------|
+| 3A              | Recommendation | Active | r0, 2021-05-07  |
+
+Authors: [@raulk], [@jhiesey], [@mxinden]
+
+Interest Group:
+
+[@raulk]: https://github.com/raulk
+[@jhiesey]: https://github.com/jhiesey
+[@mxinden]: https://github.com/mxinden
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+---
+
+## Overview
+
+The Kademlia Distributed Hash Table (DHT) subsystem in libp2p is a DHT
+implementation largely based on the Kademlia [0] whitepaper, augmented with
+notions from S/Kademlia [1], Coral [2] and the [BitTorrent DHT][bittorrent].
+
+This specification assumes the reader has prior knowledge of those systems. So
+rather than explaining DHT mechanics from scratch, we focus on differential
+areas:
+
+1. Specialisations and peculiarities of the libp2p implementation.
+2. Actual wire messages.
+3. Other algorithmic or non-standard behaviours worth pointing out.
+
+For everything else that isn't explicitly stated herein, it is safe to assume
+behaviour similar to Kademlia-based libraries.
+
+Code snippets use a Go-like syntax.
+
+## Definitions
+
+### Replication parameter (`k`)
+
+The amount of replication is governed by the replication parameter `k`. The
+default value for `k` is 20.
+
+### Kademlia routing table
+
+The data structure backing this system is a k-bucket routing table, closely
+following the design outlined in the Kademlia paper [0]. The bucket size is
+equal to the replication paramter `k`, and the maximum bucket count matches the
+size of the SHA256 function, i.e. 256 buckets.
+
+### Alpha concurrency parameter (`α`)
+
+The concurrency of node and value lookups are limited by parameter `α`, with a
+default value of 3. This implies that each lookup process can perform no more
+than 3 inflight requests, at any given time.
+
+### Distance
+
+In all cases, the distance between two keys is `XOR(sha256(key1),
+sha256(key2))`.
+
+## DHT operations
+
+The libp2p Kademlia DHT offers the following types of routing operations:
+
+- **Peer routing** - _Finding_ the closest nodes to a given key (`FIND_NODE`).
+- **Value routing** - _Putting_ a value to the nodes closest to the value's key
+  (`PUT_VALUE`) and _getting_ a value by its key from the nodes closest to that
+  key (`GET_VALUE`).
+- **Content routing** - _Adding_ oneself to the list of providers for a given
+  key at the nodes closest to that key (`ADD_PROVIDER`) and _getting_ providers
+  for a given key from the nodes closest to that key (`GET_PROVIDERS`).
+
+In addition the libp2p Kademlia DHT offers the auxiliary _bootstrap_ operation.
+
+### Peer routing
+
+The below is one possible algorithm to find nodes closest to a given key on
+the DHT. Implementations may diverge from this base algorithm as long as they
```

Comment:
"Implementations may diverge from this base algorithm": so what's the point of describing this algorithm if our implementation is different and other implementations don't have to follow this?

Reply:
Sounds good. I would just elaborate slightly, something like: "may diverge as long as they adhere to the wire format and make progress towards the target key"

Reply:
Added in https://github.com/libp2p/specs/pull/108/commits/b07409173c9efbde4b31deb87b3c30b17c0c7379.